File: TechTest/app/src/main/java/com/david_kapas/android/techtest/presentation/posts/model/PostListModel.java
package com.david_kapas.android.techtest.presentation.posts.model;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;

import com.annimon.stream.Stream;
import com.david_kapas.android.techtest.logic.api.PostsApi;
import com.david_kapas.android.techtest.logic.api.UsersApi;
import com.david_kapas.android.techtest.logic.api.model.Post;
import com.david_kapas.android.techtest.logic.api.model.PostAndUser;
import com.david_kapas.android.techtest.logic.api.model.User;
import com.david_kapas.android.techtest.logic.dao.PostDao;
import com.david_kapas.android.techtest.logic.dao.UserDao;
import com.david_kapas.android.techtest.presentation.common.model.BaseModel;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * Model class for the list of posts.
 * Created by David Kapas on 2018.03.17.
 */

public class PostListModel extends BaseModel {
    private static final String TAG = PostListModel.class.getSimpleName();

    private PostsApi postsApi;
    private PostDao postDao;
    private UsersApi usersApi;
    private UserDao userDao;

    private MutableLiveData<PostAndUserEntity> postAndUserEntityMutableLiveData = new MutableLiveData<>();
    private Observable<List<Post>> postObservable;
    private Observable<List<User>> userObservable;

    public PostListModel(PostsApi api, PostDao postDao, UsersApi usersApi, UserDao userDao) {
        this.postsApi = api;
        this.postDao = postDao;
        this.usersApi = usersApi;
        this.userDao = userDao;
    }

    public void getPosts() {
        if (postAndUserEntityMutableLiveData.getValue() == null) {
            // the proper repository pattern not implemented due to short time frame, although the model class would be more testable
            postObservable = postsApi.getPosts().subscribeOn(Schedulers.newThread())
                    .observeOn(AndroidSchedulers.mainThread())
                    .doAfterNext(this::storePostNetworkResponseInDb);
            userObservable = usersApi.getUsers().subscribeOn(Schedulers.newThread())
                    .observeOn(AndroidSchedulers.mainThread())
                    .doAfterNext(this::storeUserNetworkResponseInDb);
            getPostsAndUsers();
        }
    }

    private void getPostsAndUsers() {
        Observable.zip(postObservable, userObservable, (posts, users) -> {
            List<PostAndUser> list = new ArrayList<>();
            for (Post post : posts) {
                User currentUser = Stream.of(users).filter(user -> user.getId() == post.getUserId()).findFirst().orElse(new User());
                list.add(new PostAndUser(post, currentUser));
            }
            return list;
        }).subscribe(this::onResult, this::onError);
    }

    private void onError(Throwable throwable) {
        PostAndUserEntity postAndUserEntity = new PostAndUserEntity();
        postAndUserEntity.setError(true);
        postAndUserEntityMutableLiveData.setValue(postAndUserEntity);
    }

    private void onResult(List<PostAndUser> postAndUsers) {
        postAndUserEntityMutableLiveData.setValue(new PostAndUserEntity(postAndUsers));
    }

    public LiveData<PostAndUserEntity> getPostAndUserEntityMutableLiveData() {
        return postAndUserEntityMutableLiveData;
    }

    private void storePostNetworkResponseInDb(List<Post> posts) {
        Observable.fromCallable(() -> {
            Stream.ofNullable(posts).forEach(post -> postDao.insertPost(post));
            return true;
        }).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(this::onNext, this::onDBError);
    }

    private void storeUserNetworkResponseInDb(List<User> users) {
        Observable.fromCallable(() -> {
            Stream.ofNullable(users).forEach(user -> userDao.insertUser(user));
            return true;
        }).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(this::onNext, this::onDBError);
    }

    private void onDBError(Throwable throwable) {
        // the extended error handling not implemented due to short time frame
    }

    private void onNext(Boolean aBoolean) {

    }

}

File: TechTest/app/src/main/java/com/david_kapas/android/techtest/presentation/posts/model/PostListModel.kt
package com.david_kapas.android.techtest.presentation.posts.model

import com.annimon.stream.Stream
import com.david_kapas.android.techtest.logic.api.PostsApi
import com.david_kapas.android.techtest.logic.api.UsersApi
import com.david_kapas.android.techtest.logic.api.model.Post
import com.david_kapas.android.techtest.logic.api.model.PostAndUser
import com.david_kapas.android.techtest.logic.api.model.User
import com.david_kapas.android.techtest.logic.dao.PostDao
import com.david_kapas.android.techtest.logic.dao.UserDao
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.BiFunction
import io.reactivex.functions.Consumer
import io.reactivex.schedulers.Schedulers
import java.util.*

/**
 * Model class for the list of posts.
 * Created by David Kapas on 3/14/2019.
 */
class PostListModel(val postsApi: PostsApi, val postDao: PostDao, val usersApi: UsersApi, val userDao: UserDao) {


    var postObservable: Observable<List<Post>>? = null
    var userObservable: Observable<List<User>>? = null

    fun getPosts(): Observable<List<PostAndUser>>? {
        postObservable = postsApi.getPosts().subscribeOn(Schedulers.newThread()).observeOn(AndroidSchedulers.mainThread()).doAfterNext(this::storePostNetworkResponseInDb);
        userObservable = usersApi.getUsers().subscribeOn(Schedulers.newThread()).observeOn(AndroidSchedulers.mainThread()).doAfterNext(this::storeUserNetworkResponseInDb);
        return getPostsAndUsers();
    }


    private fun getPostsAndUsers(): Observable<List<PostAndUser>>? {
        return Observable.zip(postObservable, userObservable, BiFunction<List<Post>, List<User>, List<PostAndUser>>() { posts, users ->
            var list = ArrayList<PostAndUser>()
            for (post in posts) {
                var currentUser = Stream.of<User>(users).filter({ user -> user.id == post.userId }).findFirst().orElse(User())
                list.add(PostAndUser(post, currentUser))
            }
            list;
        });
    }

    private fun storePostNetworkResponseInDb(posts: List<Post>) {
        Single.fromCallable {
            Stream.ofNullable(posts).forEach { post -> postDao.insertPost(post) }
            true
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(Consumer<Boolean> { onNext(it) }, Consumer<Throwable> { onDBError(it) });
    }

    private fun storeUserNetworkResponseInDb(users: List<User>) {
        Single.fromCallable {
            Stream.ofNullable(users).forEach { user -> userDao.insertUser(user) }
            true
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(Consumer<Boolean> { onNext(it) }, Consumer<Throwable> { onDBError(it) })
    }

    private fun onNext(boolean: Boolean) {

    }

    private fun onDBError(throwable: Throwable) {
        // the extended error handling not implemented due to short time frame
    }
}